### ISSUE TITLE
Milestone 1: Ensure Solang works with latest Substrate

### ISSUE THREAD

**Issue by xermicus** [hyperledger/solang#969]
Solang was developed against Substrate v3.0 and older versions of both the npm polkadot api and the contract metadata crate. They need to be updated to the latest versions, and any consequent issues will need to be resolved.

The Substrate integration tests at:
https://github.com/hyperledger-labs/solang/tree/main/integration/substrate
use various polkadot npm libraries, must all be updated to the latest version, and any consequent issues must be resolved. Success for this milestone is simply that the tests are upgraded to the latestversions of all the substrate components and the tests pass, in a pull request to solang repository and any review comments are addressed.

**Status**

* [x] Update Integration tests
  * [x] Update existing test suite and use up-to-date production substrate node: #931 
  * [x] Fix any remaining regressions
  * [x] ~~Integrate and use `subxt` for integration tests: #990~~
* [x] Various bug fixes and changes in solang to better align with substrate
  * [x] #999 
  * [x] #1035
  * [x] #1011 
  * [x] #1027
  * [x] Fix anonymous values in return types
  * [x] Update calls to the `seal` API to reflect latest versions: #1025
    * [x] seal_instantiate
    * [x] seal_terminate
    * [x] seal_call (todo: flags)
    * [x] seal_set_storage
    * [x] seal_clear_storage
    * [x] seal_get_storage
    * [x] seal_tombstone_deposit (gone) #1022
    * [ ] seal_random will be removed, replace it for contract instantiation with new seal_nonce API
  * [x] ~~Implement missing `seal` API calls~~ (can be done at a later stage)
  * [x] Adapt mappings (and possibly arrays too?)to work with new storage keys
  * [x] #1045  
* [x] Re-use `ink_metadata` for ABI generation: #989 
  * [x] Allow [scale-info](https://github.com/paritytech/scale-info/pull/164) to be used at runtime
  * [x] Allow [ink_metadata](https://github.com/paritytech/ink/pull/1357) to be used at runtime
  * [x] Since the `ink!` storage refactor, `ink_metadata` requires nightly rust. We must be able to compile with stable rust (without `RUSTC_BOOTSTRAP`). [Potential fix](https://github.com/paritytech/ink/pull/1393)

**Comment by erasmus** [comment]
@xermicus thank you for the work here. Do you have an estimate of when it would be finished?

**Comment by xermicus** [comment]
> @xermicus thank you for the work here. Do you have an estimate of when it would be finished?

Very welcome. Currently blocked until the `ink!` v4 beta release, which will contain the necessary code to make this work. It is just around the corner. Based of that I expect #989 to land in our master probably next week. However, depending on your needs, you can already build solang using the branch from PR #988, to try it out right now. There might be minor changes to it but overall it already works.

**Comment by erasmus** [comment]
@xermicus great news, we will check it out 🙏

**Comment by xermicus** [comment]
@erasmus FYI #988 got merged into main. For contract development you should be good with the following:
1. Compile `solang` yourself using our `main` branch
2. [Substrate contracts node](https://github.com/paritytech/substrate-contracts-node) releases v0.19 - v.0.21 should be compatible

Please let me know whether this works for you or if you run into any problems.

There are very few broken things remaining now that I know of. The main thing is #970, so don't expect indexed (topic) events to work at all. Once this is done, we will make another solang release (probably next week) that works with the contract node versions I mentioned above and can be used until `ink!` v4 is released (this will only be next year and solang is expected to be fully compatible).

**Comment by athei** [comment]
Why is Solang not compatible with the newest contracts node (´0.22.1`)? Is this about stabilized functions changing its module (we should really stop doing that)?

**Comment by cr4pt0** [comment]
I have a solidity smart contract that I compile using Solang to substrate wasm.

```
contract Pool {
address xxx;
constructor(address _test){
  xxx = _test;
}

function callFoo(address _test) external returns (bool){
  _test.call(abi.encodeWithSignature("flip()"));
  return true;
}
```

But I received the error trying deploy the contract using polkadot js org

![image](https://user-images.githubusercontent.com/102041955/202030895-8bb555db-5640-422f-9a8c-7f673056962d.png)

I found the notes according to solang current stage

Solang works with Parity Substrate 3.0. Note that for recent Substrate Version, cross-contract calls as well as using address type as function argument or return values ARE NOT SUPPORTED. We are currently working on fixing any regressions.

https://solang.readthedocs.io/en/latest/targets/substrate.html

Does solang already support cross-contract calls as well as using address type as function argument or return values?
@athei @xermicus

**Comment by xermicus** [comment]
>  Is this about stabilized functions changing

@athei yes I thought that this [PR](https://github.com/paritytech/substrate/pull/12611) went into v0.22.1 and so it is potentially breaking, I didn't got on to check it. I'm mainly saying this here now because for one our CI currently uses v0.19, but 0.20 and 0.21 also pass the integration test, so they are safe to use. And since the front end does not yet support v0.22.1 it would not make sense to use that for anyone wanting to try it out. But I'll look into support v0.22.1 soon.

@cr4pt0 are you using solang based on our main branch? Your example works fine (metadata-wise) but you need to compile solang from our main branch yourself as the fix for this issue is not yet released [see my comment above](https://github.com/hyperledger/solang/issues/969#issuecomment-1315484644).

**Comment by cr4pt0** [comment]
@xermicus can you give example  how to
Compile solang yourself using **main** branch?

**Comment by xermicus** [comment]
@cr4pt0 yes you should do `Step 2: Build Solang`. However you don't need to build LLVM yourself, you can just use the pre-built binaries mentioned in `Step 1`.

**Comment by athei** [comment]
> @athei I had this [PR](https://github.com/paritytech/substrate/pull/12611) that went into v0.22.1 at the back of my mind and so it is potentially breaking (solang currently links some `__unstable__`), however I didn't had the chance to check it in detail.

You are right. It is annoying that functions change modules when they are stabilized. I wanna change that so that they will be added in the module they will eventually end up in if they get stabilized. They are just not there if the `unstable-interface` feature is not supplied.

**Comment by cr4pt0** [comment]
@xermicus Amazing!! Thank you so much!
